src/helper/bit-tools/line-tool.js

```diff
@@ -0,0 +1,142 @@
+import paper from '@scratch/paper';
+import {getRaster} from '../layer';
+import {forEachLinePoint, fillEllipse} from '../bitmap';
+import {getGuideLayer} from '../layer';
+import {ART_BOARD_WIDTH, ART_BOARD_HEIGHT} from '../view';
+
+/**
+ * Tool for drawing lines with the bitmap brush.
+ */
+class LineTool extends paper.Tool {
+    /**
+     * @param {!function} onUpdateSvg A callback to call when the image visibly changes
+     */
+    constructor (onUpdateSvg) {
+        super();
+        this.onUpdateSvg = onUpdateSvg;
+        
+        // We have to set these functions instead of just declaring them because
+        // paper.js tools hook up the listeners in the setter functions.
+        this.onMouseMove = this.handleMouseMove;
+        this.onMouseDown = this.handleMouseDown;
+        this.onMouseDrag = this.handleMouseDrag;
+        this.onMouseUp = this.handleMouseUp;
+
+        this.colorState = null;
+        this.active = false;
+        this.startPoint = null;
+        this.cursorPreview = null;
+        // Raster to which to draw
+        this.drawTarget = null;
+    }
+    setColor (color) {
+        this.color = color;
+    }
+    setLineSize (size) {
+        // For performance, make sure this is an integer
+        this.size = Math.max(1, ~~size);
+    }
+    // Draw a brush mark at the given point
+    draw (x, y) {
+        const roundedUpRadius = Math.ceil(this.size / 2);
+        this.drawTarget.drawImage(this.tmpCanvas, new paper.Point(~~x - roundedUpRadius, ~~y - roundedUpRadius));
+    }
+    updateCursorIfNeeded () {
+        if (!this.size) {
```

Comment:
This looks like a lot of it is copied from the bit brush tool, could you put this into a helper to avoid the copy/pasting?

Reply:
I've moved the canvas drawing part into a helper. Is that what you were looking for?